src/tool_operate.c

```diff
@@ -656,21 +656,213 @@ static void single_transfer_cleanup(struct OperationConfig *config)
+
+  CURLUcode ret = curl_url_get(uh, CURLUPART_HOST, &cid, CURLU_URLDECODE);
+  if(ret != CURLUE_OK) {
+    goto clean;
+  }
+
+  if(!cid) {
+    goto clean;
+  }
+
+  /* We might have a --ipfs-gateway argument. Check it first and use it. Error
+   * if we do have something but if it's an invalid url.
+   */
+  if(config->ipfs_gateway) {
+    if(curl_url_set(ipfsurl, CURLUPART_URL, config->ipfs_gateway, 0)
+                    == CURLUE_OK) {
+        gateway = config->ipfs_gateway;
+    }
+    else {
+      result = CURLE_BAD_FUNCTION_ARGUMENT;
+      goto clean;
+    }
+  }
+  else {
+    gateway = ipfs_gateway();
+    if(!gateway) {
+      result = CURLE_FILE_COULDNT_READ_FILE;
+      goto clean;
+    }
+
+    if(curl_url_set(ipfsurl, CURLUPART_URL, gateway, CURLU_GUESS_SCHEME
+                    | CURLU_NON_SUPPORT_SCHEME) != CURLUE_OK) {
+      goto clean;
+    }
+  }
+
+  pathbuffer = aprintf("%s/%s", protocol, cid);
+  if(!pathbuffer) {
+    goto clean;
+  }
+
+  if(curl_url_set(ipfsurl, CURLUPART_PATH, pathbuffer, CURLU_URLENCODE)
+                  != CURLUE_OK) {
+    goto clean;
+  }
+
+  if(curl_url_get(ipfsurl, CURLUPART_URL, url, CURLU_URLENCODE)
+                  != CURLUE_OK) {
+    goto clean;
+  }
+
+  if(curl_url_get(ipfsurl, CURLUPART_SCHEME, &protocol, CURLU_DEFAULT_SCHEME)
+                  != CURLUE_OK) {
+    goto clean;
+  }
+
+  result = CURLE_OK;
+
+clean:
+  if(gateway) {
+    curl_free(gateway);
+  }
+  if(cid) {
+    curl_free(cid);
+  }
+  if(pathbuffer) {
+    curl_free(pathbuffer);
+  }
+  if(ipfsurl) {
+    curl_url_cleanup(ipfsurl);
+  }
+
+  return result;
+}
+
 /*
  * Return the proto bit for the scheme used in the given URL
  */
-static long url_proto(char *url)
+static long url_proto(char **url,
+                      struct OperationConfig *config,
+                      struct GlobalConfig *global)
 {
   CURLU *uh = curl_url();
   long proto = 0;
+
   if(uh) {
-    if(url) {
-      if(!curl_url_set(uh, CURLUPART_URL, url,
-                       CURLU_GUESS_SCHEME | CURLU_NON_SUPPORT_SCHEME)) {
+    if(*url) {
+      if(curl_url_set(uh, CURLUPART_URL, *url,
+                      CURLU_GUESS_SCHEME | CURLU_NON_SUPPORT_SCHEME)
+         == CURLUE_OK) {
         char *schemep = NULL;
-        if(!curl_url_get(uh, CURLUPART_SCHEME, &schemep,
-                         CURLU_DEFAULT_SCHEME) &&
-           schemep) {
+        if(curl_url_get(uh, CURLUPART_SCHEME, &schemep,
+                         CURLU_DEFAULT_SCHEME) == CURLUE_OK
+           && schemep) {
+          if(curl_strnequal(schemep, "ipfs", 4)
+             || curl_strnequal(schemep, "ipns", 4)) {
+              CURLcode ret = ipfs_url_rewrite(uh, schemep, url, config);
+            if(ret != CURLE_OK) {
+              switch(ret) {
+              case CURLE_URL_MALFORMAT:
+                helpf(global->errors, "malformed URL. Visit https://curl.se/"
+                                      "docs/ipfs.html#Gateway-file-and-envir"
```

Comment:
I think it is better to leave out the anchor part of the URL. It seems a little fragile as that might change over time while the path is likely to remain working. That goes for the other messages too.

Another minor nit could perhaps be that we intend following lines with the open parenthesis on the line above, not with the string above.

Reply:
I do think keeping the anchor point will give a more easy way for the user to discover the right section. As there is quite some information on that page.

Is there a way to perhaps better anchor it?